### ISSUE TITLE
`reportUnusedParameter` false negative on `__new__`, `__init__` and `__init_subclass__`

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#620]
Code sample in [basedpyright playground](https://basedpyright.com/?pythonVersion=3.13&typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbABQ5AxiQIYDO9UAYmGAFzlTdQAmRwKAH0hqJDBEAKeqWAAaKADdaJAK5F2UejBABKKAFoAfFAByYFBq48bCBvWvd%2BgkZYDuUmvQXK1GrTr6xoSynDY2IEQwqiAoUF6Svuq6jnwCwqIo4kL0qgBGNPaeJN5KKuqa2nqGJuaWYeHcdoxAA)

```python
from typing import Self

class Foo:
    def __init__(self, value: str) -> None: # no error
        pass
    def __new__(cls, value: str) -> Self: # no error
        return cls(value)
    def __init_subclass__(cls, value: str) -> None: # no error
        pass
```

this is because i special-cased dunders to not report it because in many cases they should be treated the same way as an override, but these three dunders should still report it

also need to account for the fact that the `__init__` arguments need to match the `__new__` arguments, so for those i guess it should only report the error if it's unused in both?
